### ISSUE TITLE
Proposal: Remove `.extend()` and replace with `.override()`

### ISSUE THREAD

**Issue by kitsonk** [dojo/compose#84]
# Proposal

We should consider eliminating the `.extend()` API that is part of `dojo-compose` and replace it with `.override()`.  `.override()` would only allow known properties/methods on the base to be replaced and would throw if the overriding prototype contained something that was not part of the base. `.override()` would also override any mixed array properties.  For example:

``` typescript
const createFoo = compose({
        foo: [ 'bar' ]
    })
    .override({
        foo: [ 'foo' ]
    });

const foo = createFoo();
foo.foo; // [ 'foo' ]
```

Currently, extend would caused these to be "mixed in" resulting in `[ 'bar', 'foo' ]`.

We could also consider throwing on if the the underlying type is changed as well.  For example, if the target is a `string`, but a `number` is being overridden, we should throw.  When TypeScript delivers the `partial` type operator, we will also be able to guard against this at design time.
## Justification

`.extend()` promotes the concept of _traditional inheritance_ which is not the way we should think about creating classes with `dojo-compose`.  Most of the real world use cases we have for `.extend()` are actually overriding, and actually adding new functionality or changing the shape of the class is really _anti-pattern_ of how to create a maintainable class.

We are also lacking a mechanism to "overwrite" the array mixin functionality.  For example if a default class has an array of functions that provide a level of default behaviour which is not desired, the only option we have at the moment is to add another function which "undoes" what the previous function was doing.  By allowing an override, a new behaviour could be accomplished without requiring functionality to "undo" what was done earlier.

**Comment by novemberborn** [comment]
> .override() would only allow known properties/methods on the base to be replaced and would throw if the overriding prototype contained something that was not part of the base.

To clarify, "base" here is whatever `.override()` is called on? So you wouldn't be able to add new functionality, only override existing functionality?

I think that makes sense. I've gotten confused for instance with `.extend()` not taking an initializer.

**Comment by kitsonk** [comment]
We would provide both the _versions_ of the API:

``` typescript
import compose, { override } from 'dojo-compose/compose';

const createFooBar = compose({
        foo: 'foo'
    })
    .override({
        foo: 'bar'
    });

const createFooBaz = override(createFooBar, {
    foo: 'baz'
});
```

**Comment by novemberborn** [comment]
Would this solve the broken type inference in https://github.com/dojo/dataviz/pull/19?

``` ts
const createPercentageChart: ColumnChartFactory<PlayCount> = createColumnChart.extend({
    bottomAxis: {
        inputs: {
            labelSelector({ input }) { return input.artist; }
        },
        labels: { anchor: 'middle', textAnchor: 'end', rotation: -45, offset: -5 },
        ticks: { anchor: 'end', length: 10, zeroth: true }
    }
});
```

TypeScript fails to infer the `input` type (it's supposed to be `PlayCount`).

**Comment by kitsonk** [comment]
Well, actually, the first comment I would have is that is rather "surprising" construct, in that, except for arrays, any non-primative values would be shared amongst all instances of that class at creation.  That means any changes `instance.labels.anchor` would impact all other instances.

But to more specifically answer your question, no it would not fix it until TypeScript 2.1 and the `partial` keywords, since up until that point, we don't have a viable way to guard against a subtype of `createColumnChart`, but then it would basically enforce that `bottomAxis` would have to be assignable to `ColumnChart.bottomAxis`.  There seems to be a question of if `partial` will be deep or not yet.  Clearly for this use case, having a deep partial subtype would not be good, but a deep partial subtype for setting state would be.

**Comment by novemberborn** [comment]
OK let's continue that discussion in https://github.com/dojo/dataviz/pull/19#issuecomment-255671228.
